### ISSUE TITLE
Make "not flat" rules be able to work with whole object instead of only property

### ISSUE THREAD

**Issue by xepozz** [yiisoft/validator#260]
Looks logically to make `AtLeast`, `Each`, `Nested` and `Composite` rules be able to work with whole object like this:

```php
#[AtLeast(['username', 'password'])]
#[Nested(['username' => [new Required()], 'password' => [new Required()])]
#[Each([new HasLength(min: 3)])]
#[Composite([new HasLength(max: 20)])]
class Auth
{
  private $username;

  private $password;
}
```

***It's a synthetic example, it just shows how these attributes can work***

**Comment by samdark** [comment]
This looks like another way to do the same thing. For this style of declaration it's better to use rule provider and `getRules()`.

**Comment by xepozz** [comment]
> it's better to use rule provider and getRules().

it's not convenient to both with both styles of declaration. 
also here is convenient to use the Embedded rule

**Comment by samdark** [comment]
It makes no sense to me to declare validation separately from property declaration i.e. the following is way more convenient to read:

```php
class Auth
{
  #[Required()]
  #[HasLength(min: 3)]
  #[HasLength(max: 20)]
  private $username;

  #[Required()]
  #[HasLength(min: 3)]
  #[HasLength(max: 20)]
  private $password;
}
```

**Comment by xepozz** [comment]
The way that was suggested is convenient when you want to check whole object. If we talk about custom validators, it may be useful to calculate checksums, get some extra information from the validating object or use a few properties to validate, make some db queries and other.

**Comment by samdark** [comment]
1. You are specifying individual fiels or "each field" in your example. Is there a case when you aren't doing that?
2. If so, which field would receive errors?

**Comment by xepozz** [comment]
Usually the validators did this way is only used once with specific dto. All validation logic is encapsulated and it sets errors to properties inside the validator. Something like :`addError(msg, params, prop)`

**Comment by samdark** [comment]
But since it's bound to property, validation rules could be declared per property, right?

**Comment by xepozz** [comment]
They could be, as they are present now.

**Comment by vjik** [comment]
Yes, it needs to be done. This will require only a small code revision.
